Model Process CA layer hosting

eddydas opened the pull request:
#### 11061972c7486a8ddd132affabbc3396aa2ca961
<pre>
Model Process CA layer hosting
<a href="https://rdar.apple.com/123273873">rdar://123273873</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=269762">https://bugs.webkit.org/show_bug.cgi?id=269762</a>

Reviewed by Tim Horton.

Hosts a dummy green CALayer in Model process and display it for
&lt;model&gt; tag, as a precursor to actually hosting the 3D model.

The hosting mechanism utilizes the same PlatformCALayerRemoteCustom
as that used by HTMLVideoElement, with some minor generalization.

We have created a new ContentsLayerPurpose::HostedModel to
distinguish between the existing in-process Model path and the
Model process path, as a temporary measure.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::layerHostingContextIdentifier const):
(WebCore::HTMLModelElement::didUpdateLayerHostingContextIdentifier):
The hosting context identifier comes back in an async manner, so this
callback will ultimately re-create the correct hosting layer
pointing to this new hosting context identifier.

* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.cpp:
(WebCore::DummyModelPlayer::layerHostingContextIdentifier):
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.h:
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::setContentsToPlatformLayerCustom):
A generic way to utilize a PlatformLayerCustom with any
given hosting context identifier and purpose.

* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::createPlatformCALayerCustom):
(WebCore::GraphicsLayerCA::setContentsToPlatformLayerCustom):
A platformCALayerCustom essentially does one thing - hosts
another CALayer in another process. So this fits our purpose
well.

(WebCore::GraphicsLayerCA::purposeNameForInnerLayer const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
The handling for the existing &quot;Model&quot; layer purpose is quite
distinct from what we are going to do. So let&apos;s introduce
a new &quot;HostedModel&quot; layer purpose for the time being. We could
remove the old &quot;Model&quot; one once the new HostedModel implementation
is fully in parity.

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
If we have a hosting context identifier, we are ready to
create a layer that displays the hosted CA layer. This
identifier is not always available, because its hosted
CALayer is created out of process and therefore is async.
The identifier will be nullopt before the async message
comes back.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h: Renamed from Source/WebKit/ModelProcess/ModelProcessModelPlayerProxy.h.
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in: Renamed from Source/WebKit/ModelProcess/ModelProcessModelPlayerProxy.messages.in.
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm: Renamed from Source/WebKit/ModelProcess/ModelProcessModelPlayerProxy.cpp.
(WebKit::ModelProcessModelPlayerProxy::create):
(WebKit::ModelProcessModelPlayerProxy::ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::~ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::invalidate):
(WebKit::ModelProcessModelPlayerProxy::send):
(WebKit::ModelProcessModelPlayerProxy::createLayer):
(WebKit::ModelProcessModelPlayerProxy::loadModel):
(WebKit::ModelProcessModelPlayerProxy::sizeDidChange):
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.Model.sb.in:
Allows CARenderServer connection from Model process since we need to
create a hosted CALayer.

* Source/WebKit/Scripts/process-entitlements.sh:
Specify the correct sandbox profile name instead of re-using WebContent&apos;s.

* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeHostIOS.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
Now that WebProcess vends a different PlatformCALayer for Model process path,
we no longer need to guard this site. This site is now only used by
in-process path.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::~ModelProcessModelPlayer):
(WebKit::ModelProcessModelPlayer::send):
(WebKit::ModelProcessModelPlayer::didCreateLayer):
(WebKit::ModelProcessModelPlayer::load):
(WebKit::ModelProcessModelPlayer::sizeDidChange):
(WebKit::ModelProcessModelPlayer::didLoad): Deleted.
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
(WebKit::ModelProcessModelPlayer::layerHostingContextIdentifier):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.messages.in:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.cpp:
(WebKit::ModelProcessModelPlayerManager::createModelProcessModelPlayer):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:
(WebKit::GraphicsLayerCARemote::createPlatformCALayerCustom):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::create):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteCustom.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteCustom.mm:
(WebKit::PlatformCALayerRemoteCustom::create):
(WebKit::PlatformCALayerRemoteCustom::PlatformCALayerRemoteCustom):
Re-wire HTMLVideoElement&apos;s creation into the new initializer since they&apos;re
doing very much the same thing.

Canonical link: <a href="https://commits.webkit.org/275453@main">https://commits.webkit.org/275453@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dd0e9573bb827ba1454470ad72ab7be12e78a7af

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41627 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20641 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44007 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44196 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37719 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43934 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23808 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17971 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34406 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42201 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17606 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35847 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15071 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/41497 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15302 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36861 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45613 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37850 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37182 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40958 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16442 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13496 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39375 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/proxy (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18061 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9385 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18117 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17705 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -316,6 +323,15 @@ void HTMLModelElement::sizeMayHaveChanged()
+    ASSERT_UNUSED(modelPlayer, &modelPlayer == m_modelPlayer);
+    ASSERT_UNUSED(identifier, identifier.toUInt64() > 0);
+
+    if (auto* renderer = this->renderer())
```
@cdumez Per smart pointer rules, should this be a `CheckedPtr`?

hortont424 replied:
Is this going to be addressed?

eddydas replied:
Yes! Will make it:
```
    if (CheckedPtr renderer = this->renderer())
        renderer->updateFromElement();
```

pxlcoder commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -50,18 +53,47 @@ ModelProcessModelPlayerProxy::ModelProcessModelPlayerProxy(ModelProcessModelPlay
+void ModelProcessModelPlayerProxy::createLayer()
+{
+    m_layer = adoptNS([[CALayer alloc] init]);
+    [m_layer setBackgroundColor:cachedCGColor(Color::green).get()];
+    [m_layer setFrame:CGRectMake(0, 0, 100, 100)];
```
Why this color and size? Is it for debug purposes?

eddydas replied:
Mostly for debugging visualization until the whole feature is complete.
This PR deals only with CA hosting, so this is trying to put something in CA so that when we enable the model process code path, we will see model tags being rendered as green squares, proving that CA hosting from model process is working.

aprotyas commented on Source/WebCore/Modules/model-element/HTMLModelElement.h:
```diff
@@ -75,6 +76,8 @@ class HTMLModelElement final : public HTMLElement, private CachedRawResourceClie
     bool usesPlatformLayer() const;
     PlatformLayer* platformLayer() const;
 
+    std::optional<WebCore::LayerHostingContextIdentifier> layerHostingContextIdentifier() const;
```
You can skip the `WebCore::` namespace prefix since it's all in the `WebCore` namespace.

hortont424 commented on Source/WebCore/platform/graphics/GraphicsLayer.h:
```diff
@@ -539,13 +539,15 @@ class GraphicsLayer : public RefCounted<GraphicsLayer> {
     virtual void setContentsToSolidColor(const Color&) { }
     virtual void setContentsToPlatformLayer(PlatformLayer*, ContentsLayerPurpose) { }
     virtual void setContentsToPlatformLayerHost(LayerHostingContextIdentifier) { }
+    virtual void setContentsToPlatformLayerCustom(LayerHostingContextIdentifier, ContentsLayerPurpose) { }
```
Technically `setContentsToPlatformLayer` maps to the normal "Custom" PlatformCALayerRemoteCustom. This is like "setContentsToRemotePlatformContext" or something?

eddydas replied:
Good one! I'll make it `setContentsToRemotePlatformContext`

hortont424 commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -340,6 +340,12 @@ Ref<PlatformCALayer> GraphicsLayerCA::createPlatformCALayer(PlatformLayer* platf
 
+Ref<PlatformCALayer> GraphicsLayerCA::createPlatformCALayerCustom(LayerHostingContextIdentifier, PlatformCALayerClient* owner)
+{
+    ASSERT_NOT_REACHED_WITH_MESSAGE("GraphicsLayerCARemote::createPlatformCALayerCustom should always be called instead of this, but this symbol is needed to compile WebKitLegacy.");
```
This is unusual but maybe not bad?

eddydas replied:
We do have a similar assert just a few lines down there - so I figured it'd be a good thing. At the very least it helps devs to see "okay this isn't where we'd normally come across".

hortont424 commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -1190,14 +1190,18 @@ bool RenderLayerBacking::updateConfiguration(const RenderLayer* compositingAnces
         if (element->usesPlatformLayer())
             m_graphicsLayer->setContentsToPlatformLayer(element->platformLayer(), GraphicsLayer::ContentsLayerPurpose::Model);
+#if ENABLE(MODEL_PROCESS)
+        else if (auto contextID = element->layerHostingContextIdentifier(); contextID && element->document().settings().modelProcessEnabled())
```
It looks weird that ANY element with a layerHostingContextIdentifier is considered a HostedModel if modelProcessEnabled is true. Even though that's the only use for now, isn't this sort of a violation of the abstraction?

eddydas replied:
This block is under `else if (is<RenderModel>(renderer())) {` so this should be limited to a <model> element already. So within model elements, if it doesn't have a platform layer, and it has a hosting context, we treat it as a HostedModel.

Is there any other circumstances that I'm missing here?

hortont424 replied:
Nope, you're right, I misread!

hortont424 commented on Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.cpp:
```diff
@@ -61,9 +61,11 @@ ModelProcessConnection& ModelProcessModelPlayerManager::modelProcessConnection()
 Ref<ModelProcessModelPlayer> ModelProcessModelPlayerManager::createModelProcessModelPlayer(WebPage& page, WebCore::ModelPlayerClient& client)
 {
     auto identifier = WebCore::ModelPlayerIdentifier::generate();
+    // create() would send message to remote immediately, so make sure the
```
Don't think this comment is strictly necessary

hortont424 commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteCustom.mm:
```diff
@@ -61,10 +68,15 @@
+
 PlatformCALayerRemoteCustom::PlatformCALayerRemoteCustom(HTMLVideoElement& videoElement, PlatformCALayerClient* owner, RemoteLayerTreeContext& context)
-    : PlatformCALayerRemote(PlatformCALayer::LayerType::LayerTypeAVPlayerLayer, owner, context)
+    : PlatformCALayerRemoteCustom(PlatformCALayer::LayerType::LayerTypeAVPlayerLayer, videoElement.layerHostingContextID(), owner, context)
```
Double-check that video + fullscreen still work

pvollan commented on Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.Model.sb.in:
```diff
@@ -22,3 +22,6 @@
 ; THE POSSIBILITY OF SUCH DAMAGE.
 
 #include "Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in"
+
+(allow mach-lookup
+    (global-name "com.apple.CARenderServer"))
```
I think this looks good.

pvollan commented on Source/WebKit/Scripts/process-entitlements.sh:
```diff
@@ -444,7 +444,17 @@ fi
+    plistbuddy Add :com.apple.private.pac.exception bool YES
+    plistbuddy Add :com.apple.private.sandbox.profile string com.apple.WebKit.Model
+
+    webcontent_sandbox_entitlements
```
This may not be needed, since these are mostly relevant to the WebContent process.

pvollan commented on Source/WebKit/Scripts/process-entitlements.sh:
```diff
@@ -444,7 +444,17 @@ fi
+    plistbuddy Add :com.apple.QuartzCore.webkit-end-points bool YES
+    plistbuddy add :com.apple.QuartzCore.webkit-limited-types bool YES
+    plistbuddy Add :com.apple.private.memorystatus bool YES
+    plistbuddy Add :com.apple.private.webkit.use-xpc-endpoint bool YES
```
The XPC endpoint entitlement may not be needed, unless you intend to send the Launch Services database to the new process.

eddydas replied:
I don't intend so! Will remove it.

eddydas committed:
Model Process CA layer hosting
rdar://123273873
https://bugs.webkit.org/show_bug.cgi?id=269762

Reviewed by Tim Horton.

Hosts a dummy green CALayer in Model process and display it for
<model> tag, as a precursor to actually hosting the 3D model.

The hosting mechanism utilizes the same PlatformCALayerRemoteCustom
as that used by HTMLVideoElement, with some minor generalization.

We have created a new ContentsLayerPurpose::HostedModel to
distinguish between the existing in-process Model path and the
Model process path, as a temporary measure.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::layerHostingContextIdentifier const):
(WebCore::HTMLModelElement::didUpdateLayerHostingContextIdentifier):
The hosting context identifier comes back in an async manner, so this
callback will ultimately re-create the correct hosting layer
pointing to this new hosting context identifier.

* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.cpp:
(WebCore::DummyModelPlayer::layerHostingContextIdentifier):
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.h:
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::setContentsToPlatformLayerCustom):
A generic way to utilize a PlatformLayerCustom with any
given hosting context identifier and purpose.

* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::createPlatformCALayerCustom):
(WebCore::GraphicsLayerCA::setContentsToPlatformLayerCustom):
A platformCALayerCustom essentially does one thing - hosts
another CALayer in another process. So this fits our purpose
well.

(WebCore::GraphicsLayerCA::purposeNameForInnerLayer const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
The handling for the existing "Model" layer purpose is quite
distinct from what we are going to do. So let's introduce
a new "HostedModel" layer purpose for the time being. We could
remove the old "Model" one once the new HostedModel implementation
is fully in parity.

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
If we have a hosting context identifier, we are ready to
create a layer that displays the hosted CA layer. This
identifier is not always available, because its hosted
CALayer is created out of process and therefore is async.
The identifier will be nullopt before the async message
comes back.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h: Renamed from Source/WebKit/ModelProcess/ModelProcessModelPlayerProxy.h.
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in: Renamed from Source/WebKit/ModelProcess/ModelProcessModelPlayerProxy.messages.in.
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm: Renamed from Source/WebKit/ModelProcess/ModelProcessModelPlayerProxy.cpp.
(WebKit::ModelProcessModelPlayerProxy::create):
(WebKit::ModelProcessModelPlayerProxy::ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::~ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::invalidate):
(WebKit::ModelProcessModelPlayerProxy::send):
(WebKit::ModelProcessModelPlayerProxy::createLayer):
(WebKit::ModelProcessModelPlayerProxy::loadModel):
(WebKit::ModelProcessModelPlayerProxy::sizeDidChange):
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.Model.sb.in:
Allows CARenderServer connection from Model process since we need to
create a hosted CALayer.

* Source/WebKit/Scripts/process-entitlements.sh:
Specify the correct sandbox profile name instead of re-using WebContent's.

* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeHostIOS.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
Now that WebProcess vends a different PlatformCALayer for Model process path,
we no longer need to guard this site. This site is now only used by
in-process path.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::~ModelProcessModelPlayer):
(WebKit::ModelProcessModelPlayer::send):
(WebKit::ModelProcessModelPlayer::didCreateLayer):
(WebKit::ModelProcessModelPlayer::load):
(WebKit::ModelProcessModelPlayer::sizeDidChange):
(WebKit::ModelProcessModelPlayer::didLoad): Deleted.
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
(WebKit::ModelProcessModelPlayer::layerHostingContextIdentifier):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.messages.in:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.cpp:
(WebKit::ModelProcessModelPlayerManager::createModelProcessModelPlayer):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:
(WebKit::GraphicsLayerCARemote::createPlatformCALayerCustom):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::create):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteCustom.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteCustom.mm:
(WebKit::PlatformCALayerRemoteCustom::create):
(WebKit::PlatformCALayerRemoteCustom::PlatformCALayerRemoteCustom):
Re-wire HTMLVideoElement's creation into the new initializer since they're
doing very much the same thing.

Canonical link: https://commits.webkit.org/275453@main
